src/meta/src/barrier/mod.rs

```diff
@@ -162,6 +167,124 @@ pub struct GlobalBarrierManager<S: MetaStore> {
     metrics: Arc<MetaMetrics>,
 
     env: MetaSrvEnv<S>,
+
+    /// The max barrier nums in flight
+    in_flight_barrier_nums: usize,
+}
+
+struct CheckpointControl<S: MetaStore> {
+    /// Signal whether the barrier with building actor is sent
+    is_build_actor: bool,
+    /// Save the state and message of barrier in order
+    command_ctx_queue: VecDeque<EpochNode<S>>,
+}
+
+impl<S> CheckpointControl<S>
+where
+    S: MetaStore,
+{
+    fn new() -> Self {
+        Self {
+            is_build_actor: false,
+            command_ctx_queue: VecDeque::default(),
+        }
+    }
+
+    /// Return the nums of barrier (the nums of in-flight-barrier , the nums of all-barrier)
+    fn get_barrier_len(&self) -> (usize, usize) {
+        (
+            self.command_ctx_queue
+                .iter()
+                .filter(|x| matches!(x.state, InFlight))
+                .count(),
+            self.command_ctx_queue.len(),
+        )
+    }
+
+    /// Inject a `command_ctx` in `command_ctx_queue`, and it's state is `InFlight`.
+    fn inject(
+        &mut self,
+        command_ctx: Arc<CommandContext<S>>,
+        notifiers: SmallVec<[Notifier; 1]>,
+        timer: HistogramTimer,
+    ) {
+        self.command_ctx_queue.push_back(EpochNode {
+            timer: Some(timer),
+            result: None,
+            state: InFlight,
+            command_ctx: command_ctx.clone(),
+            notifiers,
+        });
+        if !matches!(command_ctx.command, Command::Plain(_)) {
+            self.is_build_actor = true;
```

Comment:
I think we need `is_build_actor`. Because the `last_node` is `complete` is not mean it is commit. So if we do not use `is_build_actor`,  we will determine each time if there is `should_pause` in the queue.

Reply:
So we may just check `last().map(|l| l.should_pause()).unwrap_or(false)`? Since...
- there should be no new commands injected after this `should_pause` command,
- currently once after this `should_pause` command is drained from the queue, we mark the `is_build_actor` to false.

Reply:
Yes , I think it should always be the last one

Reply:
Oops. Get it. Let's keep it for now.